### ISSUE TITLE
Now Dev

### ISSUE THREAD

**Issue by brunolemos** [vercel/vercel#1681]
I'm finding myself having to deploy to test things, specially when they involve redirects like an oauth flow.

I see there's #1672 for `@now/next` but I didn't find an issue tracking Now Dev in general. It would be much appreciated to have it for all build types, e.g for `@now/node` with full routing support based on the `now.json` routes.

And because I have to deploy to test some things, I've been getting `Error! You have been creating deployments at a very fast pace. Please slow down.` which blocks me from working for an unknown amount of time (seems to be at least 1h).

**Comment by paulogdm** [comment]
You are completely right @brunolemos. That's why we are developing a tool to simulate the behavior of our platform locally. This is indeed a priority tool for us and should be released very soon!
Stay tuned.

**Comment by leo** [comment]
Indeed. The `now dev` coming is coming soon!

**Comment by ckreiling** [comment]
I can't wait for this tool! I just started using Now today, and am so excited to see what this tool will look like when it becomes available.

**Comment by roychri** [comment]
Until that tool is ready, would it be possible to allow us to perform more deployments before we get asked to slowed down?
And could you tell us how long we have to wait before we can deploy again?
This is frustrating, specially when I had to do so many deployments due to error caused by isomorphic-unfetch issue!

https://spectrum.chat/zeit/general/error-500-after-deploy-success-with-zeit-now~488b2e95-63a3-4825-aa8a-47d7c37047a6

**Comment by nickerlan** [comment]
Faced the same issue. Couple of test deploys made impossible to continue.

Is it easier in paid version?

**Comment by fr3fou** [comment]
Seeing that #1672 was closed, what is going to happen? I can't think of a way to use now 2.0 without re-deploying everytime..

**Comment by brunolemos** [comment]
If you are using mainly @now/node, the current workaround is to create a file that uses micro and micro-router to create a local server with the same routes, then you can run micro-dev locally.

**Comment by pmanijak** [comment]
If you find yourself here after you wrote a bunch of lambda-capable modules for v2, and you have a working `now.json` with routes, and a `.env` file with local variables, and you're using Node, this code might save you some time putting together a local dev server:

```javascript
require('dotenv').config();

const http   = require('http');
const Router = require('router');
const serve  = require('serve-static');
const done   = require('finalhandler');
const config = require('./now.json');

// Routes
var router = Router();
config.routes.forEach(route => {
   // Ignore wildcard rules
   if (route.dest.includes("$"))
      return;

   var path = '.' + route.dest;
   const code = require(path);
   router.all(route.src, code);
});

// Static files
const www = serve('./public');

// Start server
var server = http.createServer(function (req, res) {
   // Serve routes first then fallback to static files
   router(req, res, function () {
      www(req, res, done(req, res));      
   });
});
server.listen(8080);

**Comment by fr3fou** [comment]
Seems like a viable solution, for now at least.. 

What about the future though? Is it going to stay forever as a workaround? 
This is a very specific "fix", what about other languages and deployment types?

**Comment by pmanijak** [comment]
Thanks. Those are good questions, @fr3fou. I'm just someone who ran into the same issue -- sorry I don't know the answers. Maybe all will be revealed with `now dev`.

**Comment by doprdele** [comment]
When will this be released?

**Comment by Metnew** [comment]
`micro`-based solution for routing.
- Requires `routes` being set in `now.json`
- Supports `micro-dev`
- Dynamic `require`, but it's ok for local development

Works ok so far with ~10 services (API only).

```js
const { routes } = require("./now.json");
const handlers = new Map(
	routes.map(({ src, dest }) => [src, require(`./${dest}`)])
);

module.exports = async (req, res) => {
	const dest = req.url.split("?")[0];
	const fn = handlers.get(dest) || (() => "Not supported!");
	return fn(req, res);
};
```

`routes` in `now.json`

```json
"routes": [
	{
		"src": "/crlf",
		"dest": "crlf/index.js"
	},
	{
		"src": "/subdomains",
		"dest": "subdomains/index.js"
	},
],
```

**Comment by lucasconstantino** [comment]
Just assembled some of the ideas here, and made a lib to help unit testing and locally serving Node lambdas. It respects `now.json` config, and covers most of the routing complex scenarios, such as URL rewrites. Here it goes: https://github.com/lucasconstantino/now-we-test

p.s.: Merry Christmas!

**Comment by DavidBernal** [comment]
@fr3fou totally agree

**Comment by fev4** [comment]
Constantly hitting `Error! You have been creating deployments at a very fast pace. Please slow down` at the moment. How can I avoid this since `now dev` will take a while?

Any chance the team softens this limit?

**Comment by deadmantfa** [comment]
What is the limit?

**Comment by mitevpi** [comment]
Just chiming in that I'm also running into this problem repeatedly. @fillipvt and @deadmantfa sum up my thoughts perfectly.

**Comment by JSON5995** [comment]
I get this too, be nice to be able to go as fast as is needed.

**Comment by wiiiimm** [comment]
How will non javascript endpoints be supported?
Will it be possible to run the `monorepo` demo from local?

**Comment by lucasconstantino** [comment]
@williamli It should definitely rely on Docker.

**Comment by PullJosh** [comment]
Any update on when this might be released? Even if an initial release can only include support for javascript (node, next js, etc...), having something to work with would be a huge deal.

**Comment by paulogdm** [comment]
@PullJosh Check https://github.com/zeit/now-cli/pull/1883

**Comment by TooTallNate** [comment]
The initial iteration of `now dev` has now been merged into the canary branch (https://github.com/zeit/now-cli/pull/1883#issuecomment-476953933). Gonna close this one. Thank you all for your patience.
